compiler/src/iree/compiler/Codegen/Common/ReshapePatterns.cpp

```diff
@@ -17,13 +17,47 @@
 #include "mlir/Dialect/Utils/StaticValueUtils.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 
+#define DEBUG_TYPE "iree-codegen-reshape-patterns"
+
 namespace mlir::iree_compiler {
 
 #define GEN_PASS_DEF_FOLDRESHAPEINTOINTERFACETENSORPASS
 #include "iree/compiler/Codegen/Common/Passes.h.inc"
 
 namespace {
 
+//===---------------------------------------------------------------------===//
+// Utility functions
+//===---------------------------------------------------------------------===//
+
+/// Utility to return the inner product sizes for each dimension of the
+/// provided sizes array. For example, this returns [12, 3, 1] for [2, 4, 3].
+static SmallVector<OpFoldResult> getInnerSizes(ArrayRef<OpFoldResult> sizes,
+                                               RewriterBase &rewriter) {
+  AffineExpr d0, d1;
+  bindDims(rewriter.getContext(), d0, d1);
+  AffineExpr mul = d0 * d1;
+  OpFoldResult res = rewriter.getIndexAttr(1);
+  SmallVector<OpFoldResult> innerSizes;
+  innerSizes.reserve(sizes.size());
+  for (OpFoldResult size : llvm::reverse(llvm::drop_begin(sizes))) {
+    innerSizes.push_back(res);
+    res = affine::makeComposedAffineApply(rewriter, rewriter.getUnknownLoc(),
+                                          mul, {res, size})
+              .getResult();
+  }
+  innerSizes.push_back(res);
```

Comment:
Optional: instead of handling the trailing result here, how about handling the first element outside the loop? The benefit is that we can insert `res` directly after the computation, and we don't need to think about why we compute the `res` value first but we do nothing immediately. E.g.,

```cpp
innerSizes.push_back(res);
for (OpFoldResult size : llvm::reverse(llvm::drop_begin(sizes))) {
    res = ...
    innerSizes.push_back(res);
}
```